oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/jose/JwsBuilder.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.jose;
+
+import com.nimbusds.jose.*;
+import com.nimbusds.jose.crypto.ECDSASigner;
+import com.nimbusds.jose.crypto.MACSigner;
+import com.nimbusds.jose.crypto.RSASSASigner;
+import net.minidev.json.JSONObject;
+
+import javax.crypto.SecretKey;
+import java.security.Key;
+import java.security.interfaces.ECPrivateKey;
+import java.security.interfaces.RSAPrivateKey;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * A convenience class for working with Nimbus's {@link JWSObject} API
+ *
+ * @author Josh Cummings
+ */
+public class JwsBuilder {
+	private JWSHeader.Builder header;
+	private JSONObject payload = new JSONObject();
+	private List<String> scope = new ArrayList<>();
+
+	private JWSObject jws;
+
+	public JwsBuilder(String algorithm) {
+		JWSAlgorithm alg = JWSAlgorithm.parse(algorithm);
+		this.header = new JWSHeader.Builder(alg);
+	}
+
+	/**
+	 * Add or update a field to the payload of the JWT
+	 *
+	 * @param name
+	 * @param value
+	 * @return
+	 */
+	public JwsBuilder claim(String name, Object value) {
+		this.payload.appendField(name, value);
+		return this;
+	}
+
+	/**
+	 * Set the 'exp' field on the JWT
+	 *
+	 * @param expiresAt
+	 * @return
+	 */
+	public JwsBuilder expiresAt(Instant expiresAt) {
+		this.payload.appendField("exp", expiresAt.getEpochSecond());
+		return this;
+	}
+
+	/**
+	 * Append to the 'scope' field on the JWT
+	 *
+	 * @param scope
+	 * @return
+	 */
+	public JwsBuilder scope(String scope) {
+		this.scope.add(scope);
+		return this;
+	}
+
+	/**
+	 * Sign the JWT with the first JWK in the provided JWK set
+	 *
+	 * @param builder - the JWK set to use
+	 *
+	 * @return
+	 */
+	public JwsBuilder sign(JwkSetBuilder builder) {
```

Comment:
The only thing is that if `JwkSetBuilder` has 2x `JWK` configured than arbitrarily picking the first one to sign might not work in certain test scenarios.

Reply:
Well, yes, that's true... And I agree that I'm not actually experiencing that pain since my existing tests don't configure more than one key...

...there are two reasons I see this as valuable. The first is that, when signing JWTs for real in Authorization Server, there will be this notion of picking a key based on convention (pick the most recently generated JWK, e.g.) instead of configuration (give me key "123", e.g.). The tests in this commit fall into that category of "I, the JWT, don't care which key you pick, just be consistent".

The second is that doing it this way does put a nice gloss on the tests. Consider:

```java
JwkSetBuilder good = new JwkSetBuilder().rsa("good");

String authority = new JwsBuilder("RS256").sign(good).serialize();
this.setupJwks(good);
```

vs.

```java
JwkSetBuilder good = new JwkSetBuilder().rsa("good");
Key signingKey = good.getSigningKeyById("good");

String authority = new JwsBuilder("RS256").sign("good", signingKey).serialize();
this.setupJwks(good);
```

To me, the first feels more readable, and it communicates the right semantic meaning for that test: "I don't care what you sign it with, just be consistent."

I do really like your idea, though, so, given all of that, what do you think of this:

```java
JwkSetBuilder jwks = new JwkSetBuilder().rsa("123").rsa("456");

String authority = new JwsBuilder("RS256").sign("123", jwks).serialize();
this.setupJwks(jwks);
```

I'd still add `getSigningKeyById` as you recommended, but the additional line to get the key is removed, and the `sign` method clearly indicates which key to pick.

Also, if you prefer, I can rename the `sign` method to `signWithAny`, so then we'd have:

```java
JwsBuilder#sign(String kid, JwkSetBuilder jwks); // sign with specified key
JwsBuilder#sign(String kid, Key key); // sign with specified key
JwsBuilder#signWithAny(JwkSetBuilder jwks); // sign with any valid key
```
I still prefer that they all be called `sign`, but I can see the value of having the last one named differently to hint to us that we ought to be aware of the convention (in this case, pick the first in the set).

Is there anything in the JWK spec that indicates on the signing side which JWK ought to be picked when there are multiple valid? If so then perhaps `sign(JwkSetBuilder)` could pick according to that. Seems that https://tools.ietf.org/html/rfc7517#section-5.1 says implementers can pick any order.

Reply:
Let's go with:

```
JwsBuilder#sign(String kid, JwkSetBuilder jwks);
JwsBuilder#sign(String kid, Key key);
```

I think we can leave this one out?
`JwsBuilder#signWithAny(JwkSetBuilder jwks)`

and add this:
`getSigningKeyById`

Sound good?